src/rust/grapl-observe/src/statsd_formatter.rs

```diff
@@ -0,0 +1,199 @@
+use crate::metric_error::MetricError;
+use crate::metric_error::MetricError::{MetricInvalidCharacterError, MetricInvalidSampleRateError};
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::fmt::Write;
+
+lazy_static! {
+    static ref INVALID_CHARS: Regex = Regex::new("[|#,=:]").unwrap();
+}
+
+pub struct TagPair<'a>(&'a str, &'a str);
+
+impl TagPair<'_> {
+    fn write_to_buf(&self, buf: &mut String) -> Result<(), MetricError> {
+        let TagPair(tag_key, tag_value) = self;
+        reject_invalid_chars(tag_key)?;
+        reject_invalid_chars(tag_value)?;
+        Ok(write!(buf, "{}={}", tag_key, tag_value)?)
+    }
+}
+
+fn reject_invalid_chars(s: &str) -> Result<(), MetricError> {
+    let matched = INVALID_CHARS.is_match(s);
+    if matched {
+        Err(MetricInvalidCharacterError())
+    } else {
+        Ok(())
+    }
+}
+
+pub enum MetricType {
+    Gauge,
+    Counter,
+    Millis,
+    Histogram,
+}
+
+impl MetricType {
+    fn statsd_type(&self) -> &'static str {
+        let g: &'static str = "g";
+        let c: &'static str = "c";
+        let ms: &'static str = "ms";
+        let h: &'static str = "h";
+
+        match self {
+            MetricType::Gauge => g,
+            MetricType::Counter => c,
+            MetricType::Millis => ms,
+            MetricType::Histogram => h,
+        }
+    }
+}
+
+pub fn statsd_format(
+    metric_name: &str,
+    value: f64,
+    metric_type: MetricType,
+    sample_rate: Option<f64>, // TODO: change to an `impl Into<f64>`
+    tags: &[TagPair],
+) -> Result<String, MetricError> {
+    // initial capacity chosen relatively arbitrarily
+    let mut buf: String = String::with_capacity(256);
+
+    reject_invalid_chars(metric_name)?;
+
+    write!(
+        buf,
+        "{metric_name}:{value}|{metric_type}",
+        metric_name = metric_name,
+        value = value,
+        metric_type = metric_type.statsd_type()
+    )?;
+
+    match (metric_type, sample_rate) {
+        (MetricType::Counter, Some(rate)) => {
+            // a rate of 1.0 we'll just ignore
+            if rate >= 0.0 && rate < 1.0 {
+                write!(buf, "|@{sample_rate}", sample_rate = rate)?;
+            } else {
+                return Err(MetricInvalidSampleRateError());
```

Comment:
I believe the `return` can be implicit here. I.e.
```rust
    } else {
        MetricInvalidSampleRateError()
    }
```

Reply:
I tried this, and since it's not the last thing in the function, it is not an implicit return.

I wrote and added a new test case here that would catch that.

Reply:
👍 sounds good